test/validation/api/crypto/odp_crypto_test_inp.c

```diff
@@ -342,8 +340,47 @@ static void alg_test(odp_crypto_op_t op,
 			     cipher_range,
 			     auth_range,
 			     digest_offset,
-			     is_bit_mode_cipher,
-			     is_bit_mode_auth);
+			     cipher_range_in_bits,
+			     auth_range_in_bits);
+	}
+}
+
+static void alg_test(odp_crypto_op_t op,
+		     alg_order_t order,
+		     crypto_test_reference_t *ref,
+		     odp_packet_data_range_t cipher_bit_range,
+		     odp_packet_data_range_t auth_bit_range,
+		     uint32_t digest_offset,
+		     odp_bool_t is_bit_mode_cipher,
+		     odp_bool_t is_bit_mode_auth)
+{
+	odp_packet_data_range_t cipher_range;
+	odp_packet_data_range_t auth_range;
+
+	for (int cr_in_bits = 0; cr_in_bits <= 1; cr_in_bits++) {
```

Comment:
Is it right to use 'for' loop here? Isn't it better to introduce a wrapper function like,

alg_test_flags(..., bool is_cr_in_bit, bool is_ar_in_bit)

And then have calls from here like,

alg_test_flags(...., false, false);
alg_test_flags(...., false, true);
alg_test_flags(...., true, false);
alg_test_flags(...., true, true);

Other option is to use goto,

	bool ar_in_bytes, cr_in_bytes = true;

reset_auth_test_again:
	ar_in_bytes = true;

test_again:
	cipher_range = cipher_bit_range;
	auth_range = auth_bit_range;

	if (cr_in_bytes) {
		cipher_range.offset /= 8;
		cipher_range.length /= 8;
	}

	if (ar_in_bytes) {
		auth_range.offset /= 8;
		auth_range.length /= 8;
	}

	alg_test_op_types(op, order, ref, cipher_range, auth_range,
			  digest_offset, cr_in_bits, ar_in_bits);

	if (is_bit_mode_auth && ar_in_bytes) {
		ar_in_bytes = false;
		goto test_again;
	}

	if (is_bit_mode_cipher && cr_in_bytes) {
		cr_in_bytes = false;
		goto reset_auth_test_again;
	}

If we are proceeding with 'for' loop, please check if it is all right to declare variable in for loop header.

Reply:
I do not think there is anything wrong using for statements here. We are iterating over the possible values of the two-valued conditions. I think the example using gotos for the looping is much worse. The wrapper function option would probably be ok, but it essentially just unrolls the for loops, which maybe is not so useful.

Declaring loop variables within the for statement like that has been supported since C99.

I am not planning to change this piece of code unless you insist.

Reply:
My only concern is the code legibility. You can proceed with the current code if you think that is the best way.